File: solutions/python/trees/find-largest-value-in-each-tree-row.py
# LeetCode 515 — Find Largest Value in Each Tree Row (Medium)
# Category: Trees · Approach: Level-order BFS
# Time: O(n) | Space: O(n)
# Source: https://leetcode.com/problems/find-largest-value-in-each-tree-row/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val, self.left, self.right = val, left, right

def largest_values(root: TreeNode | None) -> list[int]:
    if not root:
        return []
    res: list[int] = []
    level = [root]
    while level:
        res.append(max(node.val for node in level))
        nxt: list[TreeNode] = []
        for node in level:
            if node.left:
                nxt.append(node.left)
            if node.right:
                nxt.append(node.right)
        level = nxt
    return res

File: solutions/typescript/trees/find-largest-value-in-each-tree-row.ts
// LeetCode 515 — Find Largest Value in Each Tree Row (Medium)
// Category: Trees · Approach: Level-order BFS
// Time: O(n) | Space: O(n)
// Source: https://leetcode.com/problems/find-largest-value-in-each-tree-row/

class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;
  constructor(val = 0) {
    this.val = val;
  }
}

function largestValues(root: TreeNode | null): number[] {
  if (!root) return [];
  const res: number[] = [];
  let level: TreeNode[] = [root];
  while (level.length) {
    let mx = -Infinity;
    const next: TreeNode[] = [];
    for (const node of level) {
      if (node.val > mx) mx = node.val;
      if (node.left) next.push(node.left);
      if (node.right) next.push(node.right);
    }
    res.push(mx);
    level = next;
  }
  return res;
}
